File: src/core/request-function.ts
import type { SessionHandle } from "~/models";
import forge from "node-forge";
import pako from "pako";
import { AES } from "../api/private/aes";
import { ResponseFN } from "./response-function";
import { aesKeys } from "../api/private/keys";

/**
 * Abstraction to make requests to function API
 * of PRONOTE.
 */
export class RequestFN {
  public constructor (
    private session: SessionHandle,

    /**
     * Function name.
     *
     * This is used by the server to
     * determine the function to call.
     */
    public name: string,

    /**
     * Data given to the "secure" property.
     */
    public data: any
  ) {}

  private generate () {
    this.session.information.order++;

    const order = this.generateOrder();
    const url = new URL(`${this.session.information.url}/appelfonction/${this.session.information.accountKind}/${this.session.information.id}/${order}`);

    if (!this.session.information.skipCompression) {
      this.compress();
    }

    if (!this.session.information.skipEncryption) {
      this.encrypt();
    }

    return { order, url };
  }

  private keys () {
    return aesKeys(this.session, this.session.information.order === 1);
  }

  private generateOrder (): string {
    const { key, iv } = this.keys();
    return AES.encrypt(this.session.information.order.toString(), key, iv);
  }

  private stringify (): string {
    return forge.util.encodeUtf8(JSON.stringify(this.data) || "");
  }

  private compress (): void {
    const buffer = forge.util.createBuffer(this.stringify()).toHex();

    // We compress it using zlib, level 6, without headers.
    const deflated = pako.deflateRaw(buffer, { level: 6 });
    const bytes = Array.from(deflated).map((byte) => String.fromCharCode(byte)).join("");

    // We output it to HEX.
    this.data = forge.util.bytesToHex(bytes);
  }

  private encrypt (): void {
    const { key, iv } = this.keys();

    const data = !this.session.information.skipCompression
      // If the data has been compressed, we get the bytes from HEX.
      ? forge.util.hexToBytes(this.data)
      : this.stringify();

    this.data = AES.encrypt(data, key, iv);
  }

  public async send (): Promise<ResponseFN> {
    return this.session.queue.push(async () => {
      const payload = this.generate();

      const response = await this.session.fetcher({
        url: payload.url,
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        content: JSON.stringify({
          session: this.session.information.id,
          numeroOrdre: payload.order,
          nom: this.name,
          donneesSec: this.data
        })
      });

      return new ResponseFN(this.session, response.content);
    });
  }
}

File: library/src/jvmMain/kotlin/core/request-function.kt
package core

import api.private.AES
import api.private.Keys
import api.private.aesKeys
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import models.SessionInformation
import java.io.ByteArrayOutputStream
import java.util.zip.Deflater
import java.util.zip.DeflaterOutputStream

actual class Payload actual constructor(
    actual val order: String,
    actual val url: Url
)

actual class RequestFN actual constructor(
    private val sessionInfo: SessionInformation,
    actual val name: String,
    actual var data: String
) {
    actual fun process(): Payload {
        sessionInfo.order++

        val order = this.generateOrder()
        val url = Url("${this.sessionInfo.url}/appelfonction/${this.sessionInfo.accountKind.code}/${this.sessionInfo.id}/${order}")

        if (!this.sessionInfo.skipCompression)
            this.compress()

        if (!this.sessionInfo.skipEncryption)
            this.encrypt()

        return Payload(order, url)
    }

    private fun keys(): Keys {
        return aesKeys(this.sessionInfo, this.sessionInfo.order == 1)
    }

    private fun generateOrder(): String {
        val keys = this.keys()
        return AES.encrypt("${this.sessionInfo.order}".toByteArray(), keys.key, keys.iv)
    }

    @OptIn(ExperimentalStdlibApi::class)
    private fun compress() {
        val hexString = this.data.toByteArray().toHexString()
        val input = hexString.toByteArray(Charsets.UTF_8)

        val outputStream = ByteArrayOutputStream()
        DeflaterOutputStream(outputStream, Deflater(6, true)).use { it.write(input) }
        val output = outputStream.toByteArray()

        this.data = output.toHexString()
    }

    @OptIn(ExperimentalStdlibApi::class)
    private fun encrypt() {
        val keys = this.keys()

        if (!this.sessionInfo.skipCompression)
            this.data = AES.encrypt(this.data.hexToByteArray(), keys.key, keys.iv)
        else
            this.data = AES.encrypt(this.data.toByteArray(), keys.key, keys.iv)
    }

    actual suspend fun send(): ResponseFN {
        val payload = this.process()

        val requestData = buildJsonObject {
            put("session", sessionInfo.id)
            put("numeroOrdre", payload.order)
            put("nom", name)
            put("donneesSec", Json.parseToJsonElement(data))
        }

        val response = HttpClient().request(payload.url) {
            method = HttpMethod.Post
            header("Content-Type", "application/json")
            setBody(Json.encodeToString(requestData))
        }

        return ResponseFN(this.sessionInfo, response.bodyAsText())
    }
}